### PR TITLE
[workspace cluster] Rework start cluster selection

### LINKED ISSUE #7947: Implement https://www.notion.so/gitpod/Workspace-start-cluster-selection-9dceafb290214af891cb31521ca8678d

### PR DESCRIPTION
**Pull request by csweichel**
## Description
This PR changes the workspace cluster selection during workspace startup to match the [Workspace start cluster selection RFC](https://www.notion.so/gitpod/Workspace-start-cluster-selection-9dceafb290214af891cb31521ca8678d). It also introduces a new `has-more-resources` admission constraint which is tied to the subscription of a user.

## Related Issue(s)

Fixes #7947

## How to test
- ensure the unit test additions cover enough ground
- TBD

## Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **csweichel** and additionally assign **geropl** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @geropl` in a comment when ready.

Associated issue: *#7947*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[components/ws-manager-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-api/OWNERS)**
- **[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)**
- **[components/ws-manager-bridge-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge-api/OWNERS)**
- **[dev/gpctl/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/dev/gpctl/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/gitpod-db/src/typeorm/migration/1643724196672-RemoveAdmissionPreferences.ts
Comment (geropl) [inline components/gitpod-db/src/typeorm/migration/1643724196672-RemoveAdmissionPreferences.ts:15]: `DRP{}` typo

File: components/server/src/workspace/workspace-starter.ts
Comment (geropl) [inline components/server/src/workspace/workspace-starter.ts:187]: Why not `for (const cluster of clusters)`

File: components/server/src/workspace/workspace-starter.ts
Comment (geropl) [inline components/server/src/workspace/workspace-starter.ts:189]: the comment is outdated now.

File: components/server/src/workspace/workspace-starter.ts
Comment (geropl) [reply components/server/src/workspace/workspace-starter.ts:187]: I see, an (async) iterator. Long time not seen. :-D

Comment (csweichel) [reply components/server/src/workspace/workspace-starter.ts:187]: `clusters` is an async iterator, hence requiring the `await`

File: components/gitpod-db/src/typeorm/migration/1643724196672-RemoveAdmissionPreferences.ts
Comment (csweichel) [reply components/gitpod-db/src/typeorm/migration/1643724196672-RemoveAdmissionPreferences.ts]: fixed

File: components/ws-manager-api/typescript/src/constraints.ts
Comment (geropl) [inline components/ws-manager-api/typescript/src/constraints.ts:85]: This (`s.includes(c)`) only works as long as contraints only filter, but never manipulate shapes. I think that assumption is reasonable.

File: components/ws-manager-api/typescript/src/client-provider.ts
Comment (geropl) [inline components/ws-manager-api/typescript/src/client-provider.ts:57]: I find the asyncIterator syntax and usage rather noisy and hard to read/follow, mostly because it's so uncommon.
I assume it's used here to hide the lazy initialization of clients/filtering of used clients here. Not sure I would have gone the same path, but fine with me as is.

File: components/ws-manager-bridge/src/cluster-service-server.ts
Comment (geropl) [inline components/ws-manager-bridge/src/cluster-service-server.ts:200]: :+1: Really like this going away. Having just `admissionContraints` is IMO easier to wrap yur head around.

**Comment by geropl** [comment]
Testing now... devstaging, at least.

**Review by geropl — APPROVED** [review]
The code changes - and especially the tests - LGTM

During testing I could see that the relevant code seemed to work in devstaging, alltough the workspace did not come up due to this (unrelated):
image

I will take a note that we double-check staging before we deploy tomorrow morning, just to test the cross-cluster stuff. Just to be on the save side here. :slightly_smiling_face:

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: f5332fa30fc59fbd4f3513826e5edace95d0d12e</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7949#pullrequestreview-870763037" title="Approved">geropl</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **csweichel** after the PR has been reviewed.

Associated issue: *#7947*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl]
- **[components/ws-manager-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-api/OWNERS)**
- ~~[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)~~ [geropl]
- ~~[components/ws-manager-bridge-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge-api/OWNERS)~~ [geropl]
- **[dev/gpctl/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/dev/gpctl/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/server/ee/src/user/eligibility-service.ts
Comment (sagor999) [inline components/server/ee/src/user/eligibility-service.ts:261]: only PRO plans are eligible? or is it PRO and everything above?

**Review by sagor999 — APPROVED** [review]

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7949#pullrequestreview-870763037" title="Approved">geropl</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7949#pullrequestreview-870834360" title="Approved">sagor999</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **csweichel** after the PR has been reviewed.

Associated issue: *#7947*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl]
- **[components/ws-manager-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-api/OWNERS)**
- ~~[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)~~ [geropl]
- ~~[components/ws-manager-bridge-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge-api/OWNERS)~~ [geropl]
- ~~[dev/gpctl/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/dev/gpctl/OWNERS)~~ [sagor999]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/ws-manager-api/typescript/src/constraints.ts
Comment (csweichel) [reply components/ws-manager-api/typescript/src/constraints.ts:24]: aye, that would have been great - I don't want to reset the LGTM label now though. I'll add that in in a follow up.

File: components/server/ee/src/user/eligibility-service.ts
Comment (csweichel) [reply components/server/ee/src/user/eligibility-service.ts:261]: Indeed, for now this is only the unleashed/pro plans. We might have to revisit this depending on the plans of customers affected by this change.

File: components/ws-manager-api/typescript/src/client-provider.ts
Comment (csweichel) [reply components/ws-manager-api/typescript/src/client-provider.ts:57]: Same here - it's a rather unsightly construct. Originally I used a custom "pop" function, but thought that moving to something more common is the way to go. Not sure that was worth the trouble though.

The purpose is indeed to facilitate lazy init of the clients.

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7949#issuecomment-1028167884" title="Approved">csweichel</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7949#pullrequestreview-870763037" title="Approved">geropl</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7949#pullrequestreview-870834360" title="Approved">sagor999</a>*

Associated issue: *#7947*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl]
- ~~[components/ws-manager-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-api/OWNERS)~~ [csweichel]
- ~~[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)~~ [geropl]
- ~~[components/ws-manager-bridge-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge-api/OWNERS)~~ [csweichel,geropl]
- ~~[dev/gpctl/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/dev/gpctl/OWNERS)~~ [csweichel,sagor999]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/ws-manager-api/typescript/src/constraints.ts
Comment (csweichel) [reply components/ws-manager-api/typescript/src/constraints.ts:24]: done

**Comment by roboquat** [comment]
New changes are detected. LGTM label has been removed.

**Review by sagor999 — APPROVED** [review]

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 4d97cfc9f4259e403850620bf50ca19d0f149449</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/7949#issuecomment-1028167884" title="Approved">csweichel</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7949#pullrequestreview-870763037" title="Approved">geropl</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/7949#pullrequestreview-870903668" title="Approved">sagor999</a>*

Associated issue: *#7947*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [geropl]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [geropl]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [geropl]
- ~~[components/ws-manager-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-api/OWNERS)~~ [csweichel]
- ~~[components/ws-manager-bridge/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge/OWNERS)~~ [geropl]
- ~~[components/ws-manager-bridge-api/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/ws-manager-bridge-api/OWNERS)~~ [csweichel,geropl]
- ~~[dev/gpctl/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/dev/gpctl/OWNERS)~~ [csweichel,sagor999]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
